### DISCUSSION TITLE
listen tcp4 0.0.0.0:9999: bind: address already in use

### DISCUSSION THREAD

**Discussion by mohammadih** [srsran/srsRAN_Project#616] [General Help]
when I want to build the docker, if I stop the webui it works but not have access to define the UEs information like Imsi, if I restart the webui again the docekr won't work since the port is busy, is there anyway to change this port address or fix this problem?
this is the error i get:

 ✔ Container open5gs_5gc  Recreated                                                                                                                                                                    0.1s 
Attaching to open5gs_5gc
Error response from daemon: driver failed programming external connectivity on endpoint open5gs_5gc (7f4fa0ac35278edbddfc0b44d7cf84f9f780bff226b74240a7e2f09ee9aee85e): Error starting userland proxy: listen tcp4 0.0.0.0:9999: bind: address already in use

**Comment by pgawlowicz** [comment]
Are you sure there is no there other process using port=9999? do you stop the first open5gs container?

**Reply by mohammadih** [reply]
yes, I checked, the problem is each time I kill the port, it restart automatically

**Reply by pgawlowicz** [reply]
which process is running on 9999?

**Reply by mohammadih** [reply]
take a look at screenshot, the localhost:9999 
each time i kill the port, it restart imidiately with another PID
![docker_connection_problem](https://github.com/srsran/srsRAN_Project/assets/60555990/035cb877-11a8-40da-a9de-1c22d868c965)

**Reply by chelimi** [reply]
I'm also getting same error.please solve.

waxwing@waxwing:~/srsran/srsRAN_Project/docker$ sudo lsof -i -P -n | grep 9999
node      9269            root   29u  IPv4 108888      0t0  TCP 127.0.0.1:9999 (LISTEN)
waxwing@waxwing:~/srsran/srsRAN_Project/docker$ sudo kill -9 9269
waxwing@waxwing:~/srsran/srsRAN_Project/docker$ sudo docker-compose up --build 5gc
[+] Building 2.5s (16/16) FINISHED                                                                                                                                                           docker:default
 => [5gc internal] load .dockerignore                                                                                                                                                     0.2s
 => => transferring context: 2B                                                                                                                                                              0.0s
 => [5gc internal] load build definition from Dockerfile                                                                                                                           0.3s
 => => transferring dockerfile: 3.11kB                                                                                                                                                  0.0s
 => [5gc internal] load metadata for docker.io/library/ubuntu:22.04                                                                                                      1.1s
 => [5gc internal] load build context                                                                                                                                                       0.2s
 => => transferring context: 142B                                                                                                                                                          0.0s
 => [5gc base 1/8] FROM docker.io/library/ubuntu:22.04@sha256:a6d2b38300ce017add71440577d5b0a90460d0e57fd7aec21dd0d1b0761bbfb2                                                                         0.0s
 => CACHED [5gc base 2/8] RUN DEBIAN_FRONTEND=noninteractive apt-get update     && apt install -y software-properties-common     && rm -rf /var/lib/apt/lists/*                                        0.0s
 => CACHED [5gc base 3/8] RUN DEBIAN_FRONTEND=noninteractive apt-get update     && apt-get install -y     python3-pip     python3-setuptools     python3-wheel     ninja-build     build-essential     0.0s
 => CACHED [5gc base 4/8] RUN DEBIAN_FRONTEND=noninteractive apt-get update && apt-get install -y --no-install-recommends wget gnupg     && wget -qO - https://www.mongodb.org/static/pgp/server-6.0.  0.0s
 => CACHED [5gc base 5/8] RUN echo v2.7.0 > ./open5gsversion                                                                                                                                           0.0s
 => CACHED [5gc base 6/8] RUN git clone --depth 1 --branch $(cat ./open5gsversion) https://github.com/open5gs/open5gs open5gs    && cd open5gs     && meson build --prefix=`pwd`/install     && ninja  0.0s
 => CACHED [5gc base 7/8] RUN curl -fsSL https://deb.nodesource.com/setup_20.x | bash -     && apt-get install -y nodejs     && cd open5gs/webui     && npm ci --no-optional                           0.0s
 => CACHED [5gc base 8/8] RUN python3 -m pip install pymongo click pyroute2 ipaddress python-iptables                                                                                                  0.0s
 => CACHED [5gc open5gs 1/3] WORKDIR /open5gs                                                                                                                                                          0.0s
 => CACHED [5gc open5gs 2/3] COPY open5gs-5gc.yml open5gs-5gc.yml.in                                                                                                                                   0.0s
 => CACHED [5gc open5gs 3/3] COPY open5gs_entrypoint.sh add_users.py setup_tun.py subscriber_db.cs[v] ./                                                                                               0.0s
 => [5gc] exporting to image                                                                                                                                                                           0.1s
 => => exporting layers                                                                                                                                                                                0.0s
 => => writing image sha256:08070cfd35e4496cf9aacafe920e3eac994d45af1a598e5ac6d2669d7afb59ae                                                                                                           0.0s
 => => naming to docker.io/library/docker-5gc                                                                                                                                                          0.0s
Attaching to open5gs_5gc
Error response from daemon: driver failed programming external connectivity on endpoint open5gs_5gc (e466aa0d92d6fe1ad2f9ff575ad2fe1ca74902aa262e31032998218ea4b034c5): Error starting userland proxy: listen tcp4 0.0.0.0:9999: bind: address already in use

**Reply by mohammadih** [reply]
I also open another discussion on opeg5gs discussion as well, let's see what is the answer. #3213
you can post your question there as well. it might catch their attention to our problem

**Reply by pgawlowicz** [reply]
you can try to change port used on the host machine in this line:
https://github.com/srsran/srsRAN_Project/blob/main/docker/docker-compose.yml#L36

currently, it is exposing port 9999 from the container as port 9999 on the host machine

**Reply by chelimi** [reply]
after solving that issue im getting another one 
waxwing@waxwing:~/srsran/srsRAN_Project/build$ sudo gnb -c gnb_zmq.yaml 
[sudo] password for waxwing: 

The PRACH detector will not meet the performance requirements with the configuration {Format 0, ZCZ 0, SCS 1.25kHz, Rx ports 1}.
Lower PHY in executor blocking mode.

--== srsRAN gNB (commit 0b2702cca) ==--

Connecting to AMF on 10.53.1.2:38412
Available radio types: uhd.
Factory for radio type zmq not found. Make sure to select a valid type.
srsGNB ERROR: Unable to create radio session.




I again installed zmq libraries and srsran_project
Please solve this issue

**Reply by pgawlowicz** [reply]
please do
```
git clone https://github.com/srsran/srsRAN_Project.git
cd srsRAN_Project
mkdir build
cd build
cmake ../ -DENABLE_EXPORT=ON -DENABLE_ZEROMQ=ON
```
and share the terminal output

**Reply by chelimi** [reply]
hub.com/srsran/srsRAN_Project.git
Cloning into 'srsRAN_Project'...
remote: Enumerating objects: 177069, done.
remote: Counting objects: 100% (18074/18074), done.
remote: Compressing objects: 100% (3897/3897), done.
remote: Total 177069 (delta 14549), reused 16638 (delta 13973), pack-reused 158995
Receiving objects: 100% (177069/177069), 51.22 MiB | 3.06 MiB/s, done.
Resolving deltas: 100% (144370/144370), done.
waxwing@waxwing:~/srsran$ cd srsRAN_Project
waxwing@waxwing:~/srsran/srsRAN_Project$ mkdir build
waxwing@waxwing:~/srsran/srsRAN_Project$ cd build
waxwing@waxwing:~/srsran/srsRAN_Project/build$ cmake ../ -DENABLE_EXPORT=ON -DENABLE_ZEROMQ=ON
-- The C compiler identification is GNU 11.4.0
-- The CXX compiler identification is GNU 11.4.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Build type not specified: defaulting to Release.
-- Performing Test HAS_MAYBE_UNINITIALIZED
-- Performing Test HAS_MAYBE_UNINITIALIZED - Success
-- Performing Test HAVE_NON_VIRTUAL_DTOR
-- Performing Test HAVE_NON_VIRTUAL_DTOR - Success
-- Performing Test HAVE_SUGGEST_OVERRIDE
-- Performing Test HAVE_SUGGEST_OVERRIDE - Success
-- Performing Test HAVE_SHADOW
-- Performing Test HAVE_SHADOW - Success
-- Assertion level set to NORMAL
-- Could NOT find libdw (missing: LIBDW_LIBRARY LIBDW_INCLUDE_DIR) 
-- Could NOT find libbfd (missing: LIBBFD_LIBRARY LIBBFD_INCLUDE_DIR) 
-- Could NOT find libdwarf (missing: LIBDWARF_LIBRARY LIBDWARF_INCLUDE_DIR LIBELF_LIBRARY LIBELF_INCLUDE_DIR) 
-- Found Backward: /home/waxwing/srsran/srsRAN_Project/cmake/modules  
-- Backward-cpp found, but external libraries are missing.
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.2") 
-- Checking for module 'mbedtls'
--   No package 'mbedtls' found
-- MBEDTLS LIBRARIES: /usr/lib/x86_64-linux-gnu/libmbedcrypto.so
-- MBEDTLS STATIC LIBRARIES: /usr/lib/x86_64-linux-gnu/libmbedcrypto.a
-- MBEDTLS INCLUDE DIRS: /usr/include
-- Found MbedTLS: /usr/lib/x86_64-linux-gnu/libmbedcrypto.so  
-- Checking for module 'fftw3f >= 3.0'
--   Found fftw3f , version 3.3.8
-- FFTW3F LIBRARIES: /usr/lib/x86_64-linux-gnu/libfftw3f.so
-- FFTW3F STATIC LIBRARIES: /usr/lib/x86_64-linux-gnu/libfftw3f.a
-- FFTW3F INCLUDE DIRS: /usr/include
-- Found FFTW3F: /usr/lib/x86_64-linux-gnu/libfftw3f.so  
-- Found GTest: /usr/lib/x86_64-linux-gnu/cmake/GTest/GTestConfig.cmake (found version "1.11.0")  
-- Found Threads: TRUE  
-- UHD LIBRARIES /usr/local/lib/libuhd.so
-- UHD INCLUDE DIRS /usr/local/include
-- Found UHD: /usr/local/lib/libuhd.so  
-- Checking for module 'yaml-cpp'
--   Found yaml-cpp, version 0.7.0
-- YAMLCPP LIBRARIES: /usr/lib/x86_64-linux-gnu/libyaml-cpp.so
-- YAMLCPP INCLUDE DIRS: /usr/include
-- Found YAMLCPP: /usr/lib/x86_64-linux-gnu/libyaml-cpp.so  
-- FINDING ZEROMQ.
-- Checking for module 'ZeroMQ'
--   No package 'ZeroMQ' found
-- Found libZEROMQ: /usr/local/include, /usr/local/lib/libzmq.so
-- Performing Test HAVE_SSE
-- Performing Test HAVE_SSE - Success
-- SSE4.1 is enabled - target CPU must support it
-- Performing Test HAVE_AVX
-- Performing Test HAVE_AVX - Success
-- AVX is enabled - target CPU must support it
-- Performing Test HAVE_AVX2
-- Performing Test HAVE_AVX2 - Success
-- AVX2 is enabled - target CPU must support it
-- Performing Test HAVE_FMA
-- Performing Test HAVE_FMA - Success
-- FMA is enabled - target CPU must support it
-- Performing Test HAVE_AVX512
-- Performing Test HAVE_AVX512 - Failed
-- Performing Test HAVE_MARCH
-- Performing Test HAVE_MARCH - Success
-- Enabling ccache for C
-- Enabling ccache for CXX
-- Found Doxygen: /usr/bin/doxygen (found version "1.9.1") found components: doxygen dot 
-- Checking for module 'sctp'
--   No package 'sctp' found
-- SCTP LIBRARIES: /usr/lib/x86_64-linux-gnu/libsctp.so
-- SCTP INCLUDE DIRS: /usr/include
-- Found SCTP: /usr/lib/x86_64-linux-gnu/libsctp.so  
-- FFT_LIBRARIES: /usr/lib/x86_64-linux-gnu/libfftw3f.so
-- Building srsRAN version 24.04.0
-- Configuring done
-- Generating done
-- Build files have been written to: /home/waxwing/srsran/srsRAN_Project/build
waxwing@waxwing:~/srsran/srsRAN_Project/build$

**Reply by pgawlowicz** [reply]
ok, zmq is found, please compile the project again and check whether it works.

**Reply by chelimi** [reply]
its running but not connecting to ue
waxwing@waxwing:~/srsran/srsRAN_Project/build/apps/gnb$ sudo gnb -c gnb_zmq.yaml 

The PRACH detector will not meet the performance requirements with the configuration {Format 0, ZCZ 0, SCS 1.25kHz, Rx ports 1}.
Lower PHY in executor blocking mode.

--== srsRAN gNB (commit 78238fd15) ==--

Connecting to AMF on 10.53.1.2:38412
Available radio types: uhd and zmq.
Cell pci=1, bw=20 MHz, 1T1R, dl_arfcn=368500 (n3), dl_freq=1842.5 MHz, dl_ssb_arfcn=368410, ul_freq=1747.5 MHz

==== gNodeB started ===


waxwing@waxwing:~$ sudo srsue
Active RF plugins: libsrsran_rf_uhd.so libsrsran_rf_zmq.so
Inactive RF plugins: 
Couldn't open , trying /root/.config/srsran/ue.conf
Reading configuration file /root/.config/srsran/ue.conf...

Built in Release mode using commit ce8a3cae1 on branch HEAD.

Opening 1 channels in RF device=zmq with args=tx_port=tcp://*:2001,rx_port=tcp://localhost:2000,id=ue,base_srate=11.52e6
Supported RF device list: UHD zmq file
CHx base_srate=11.52e6
CHx id=ue
Current sample rate is 1.92 MHz with a base rate of 11.52 MHz (x6 decimation)
CH0 rx_port=tcp://localhost:2000
CH0 tx_port=tcp://*:2001
Waiting PHY to initialize ... done!
Attaching UE...
Current sample rate is 11.52 MHz with a base rate of 11.52 MHz (x1 decimation)
Current sample rate is 11.52 MHz with a base rate of 11.52 MHz (x1 decimation)
.................

**Reply by pgawlowicz** [reply]
You need to start the srsue with a config file. please follow this tutorial: https://docs.srsran.com/projects/project/en/latest/tutorials/source/srsUE/source/index.html#zeromq-based-setup

**Reply by mohammadih** [reply]
I tried 8888:9999 before, but now after changing the port number in docker-compose to 8888:8888, problem solved, however the localhost still is http://localhost:9999.
Thanks

**Reply by pgawlowicz** [reply]
do you have a manually installed open5gs version running on the same PC?

**Reply by pgawlowicz** [reply]
@mohammadih any update on this issue?
